osu.Game/Screens/Import/ReplayMissingBeatmapScreen.cs

```diff
@@ -0,0 +1,225 @@
+                                    Spacing = new Vector2(15),
+                                    Children = new Drawable[]
+                                    {
+                                        beatmapPanelContainer = new Container
+                                        {
+                                            AutoSizeAxes = Axes.Both,
+                                            Anchor = Anchor.CentreLeft,
+                                            Origin = Anchor.CentreLeft,
+                                        },
+                                    }
+                                },
+                                automaticDownload = new SettingsCheckbox
+                                {
+                                    LabelText = "Automatically download beatmaps",
+                                    Current = config.GetBindable<bool>(OsuSetting.AutomaticallyDownloadWhenSpectating),
+                                    Anchor = Anchor.Centre,
+                                    Origin = Anchor.Centre,
+                                },
+                                replayDownloadButton = new ReplayDownloadButton(new ScoreInfo())
+                                {
+                                    Anchor = Anchor.Centre,
+                                    Origin = Anchor.Centre,
+                                },
+                            }
+                        }
+                    }
+                },
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            if (beatmapSetInfo == null)
+            {
+                var onlineBeatmapRequest = new GetBeatmapSetRequest(beatmap.OnlineBeatmapSetID);
+
+                onlineBeatmapRequest.Success += res =>
+                {
+                    beatmapSetInfo = res;
+                    updateStatus();
+                };
+                api.Queue(onlineBeatmapRequest);
+            }
+
+            updateStatus();
+            realmSubscription = realm.RegisterForNotifications(
+                realm => realm.All<BeatmapSetInfo>().Where(s => !s.DeletePending), beatmapsChanged);
+        }
+
+        private void updateStatus()
+        {
+            if (beatmapSetInfo == null) return;
+
+            beatmapPanelContainer.Clear();
+            beatmapPanelContainer.Child = new BeatmapCardNormal(beatmapSetInfo, allowExpansion: false);
+            checkForAutomaticDownload(beatmapSetInfo);
+        }
+
+        private void checkForAutomaticDownload(APIBeatmapSet beatmap)
+        {
+            if (!automaticDownload.Current.Value)
+                return;
+
+            beatmapDownloader.Download(beatmap);
+        }
+
+        private void beatmapsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeSet? changes)
+        {
+            if (changes?.InsertedIndices == null) return;
+
+            if (beatmapSetInfo == null) return;
```

Comment:
Does this mean what I'm thinking it means, namely that potentially a beatmap import can be missed entirely if the subscription completes before the API request for the beatmap set does?

Reply:
You misunderstand. The scenario I'm attempting to describe is:

```
API thread			import thread
----------------------------------------------------------------
begins beatmap set fetch
				begins importing beatmsp set
				to local database
				
				finishes importing beatmap set
				to local database
				
				realm subscription fires
				
				beatmap set hasn't yet been
				fetched from online
				=> subscription does nothing

beatmap set fetched
```

But as [already stated above](https://github.com/ppy/osu/pull/24450#discussion_r1287690508) I'm 99% sure all of this "nullable beatmap set" stuff can just be removed with no repercussions.

Reply:
you mean declare that APIBeatmap.BeatmapSet is not null?

Reply:
Yes.